### PR TITLE
fix: intro preauth client once

### PR DESCRIPTION
**Pull request by pike1212**
## Related issue

#712 

## Proposed changes

Only initializes oauth client once during config stage per config and doesn't initialize when creating new introspector.

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

File: pipeline/authn/authenticator_oauth2_introspection.go [hunk 1]
```diff
@@ -252,61 +252,66 @@ func (a *AuthenticatorOAuth2Introspection) Validate(config json.RawMessage) erro
...
+func (a *AuthenticatorOAuth2Introspection) Config(config json.RawMessage) (*AuthenticatorOAuth2IntrospectionConfiguration, *http.Client, error) {
 	var c AuthenticatorOAuth2IntrospectionConfiguration
 	if err := a.c.AuthenticatorConfig(a.GetID(), config, &c); err != nil {
-		return nil, NewErrAuthenticatorMisconfigured(a, err)
+		return nil, nil, NewErrAuthenticatorMisconfigured(a, err)
 	}
 
-	var rt http.RoundTripper
+	clientKey := fmt.Sprintf("%x", md5.Sum([]byte(config)))
+	client, ok := a.clientMap[clientKey]
```
[the remark this answers is not in the document]
Comment (pike1212) [reply pipeline/authn/authenticator_oauth2_introspection.go:269]: TIL, let me know how that looks

File: pipeline/authn/authenticator_oauth2_introspection.go [hunk 2]
```diff
@@ -252,61 +254,71 @@ func (a *AuthenticatorOAuth2Introspection) Validate(config json.RawMessage) erro
...
 
-	var rt http.RoundTripper
+	clientKey := fmt.Sprintf("%x", md5.Sum([]byte(config)))
+	a.mu.RLock()
+	client, ok := a.clientMap[clientKey]
+	a.mu.RUnlock()
 
-	if c.PreAuth != nil && c.PreAuth.Enabled {
-		var ep url.Values
+	if !ok {
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection.go:272]: Should we add a test for this? I imagine a test which creates one instance for `AuthenticatorOAuth2Introspection`, uses three configs (one with preauth, one without preauth, one with another preauth) and then calls `Config` and expecnts the first config to always produce the same client (checking the pointer), the second never returns a client, and the third returns a different client. To make order random I would probably call config 1, 2, 3 - and then 2, 3, 1 to ensure this does not have side effects.

Comment (pike1212) [reply pipeline/authn/authenticator_oauth2_introspection.go:272]: I added a test, let me know what you think...

File: pipeline/authn/authenticator_oauth2_introspection_test.go [hunk 3]
```diff
@@ -589,4 +590,61 @@ func TestAuthenticatorOAuth2Introspection(t *testing.T) {
...
+		if err != nil {
+			require.NoError(t, err)
+		}
+
+		_, preauthTwoClient, err := authenticator.Config(preAuthConfigTwo)
+		if err != nil {
+			require.NoError(t, err)
+		}
+
+		if noPreauthClient == preauthOneClient || noPreauthClient == preauthTwoClient || preauthOneClient == preauthTwoClient {
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection_test.go:617]: You could simplify this with:

```suggestion
		require.NotEqual(t, noPreauthClient, preauthOneClient )
		// ...
```

File: pipeline/authn/authenticator_oauth2_introspection_test.go [hunk 4]
```diff
@@ -589,4 +590,61 @@ func TestAuthenticatorOAuth2Introspection(t *testing.T) {
...
+		if err != nil {
+			require.NoError(t, err)
+		}
+		if noPreauthClient2 != noPreauthClient {
+			t.FailNow()
+		}
+
+		if noPreauthClient == preauthOneClient || noPreauthClient == preauthTwoClient || preauthOneClient == preauthTwoClient {
+			t.FailNow()
+		}
```
Comment (aeneasr) [inline pipeline/authn/authenticator_oauth2_introspection_test.go:647]: Same here :)

**Review by aeneasr — COMMENTED** [review]
Awesome! I think we can use require/assert for the equals comparison but otherwise this is exactly what I imagined! thank you :)

**Comment by pike1212** [comment]
@aeneasr can this be merged?

**Comment by aeneasr** [comment]
Yes! :)

**Comment by aeneasr** [comment]
Thank you! :)

**Pull request merged**